### PR TITLE
postprocessing_gateways

### PR DESCRIPTION
**Pull request by leclem**
quality controls

cutting video at the right time work

examples

script_example.py

Better quality function

changes working

work working

commit

solving bugs and code reviews

sonarcloud corrections

test_quality_check added

removing prints

removing url

generating zoom video in case regeneration fails

bug lower

new webp format

### CODE REVIEW

**Comment by leclem** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/160?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#163** <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/163?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#162** <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/162?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#161** <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/161?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#160** <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/160?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/160?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#159** <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/159?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#158** <a href="https://app.graphite.dev/github/pr/Vikit-ai/sdk/158?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: vikit/gateways/vikit_gateway.py [hunk 1]
```diff
@@ -940,13 +940,14 @@ async def generate_video_from_image_and_text_runway(self, prompt, aspect_ratio=(
...
-                    image_data = base64.b64decode(prompt.image)
-                except Exception: 
+                
+                if prompt.image.split('.') is not None and prompt.image.split('.')[-1].lower() in ["jpg", "png", "jpeg", "bmp", "tiff", "webp"]:
                     #Read file path and then convert to Base64
                     with open(prompt.image, "rb") as image_file:
                         image_data = image_file.read()
+                else: 
+                    # Convert result to Base64
+                    image_data = base64.b64decode(prompt.image)
```
Comment (jeffmac-aix) [inline vikit/gateways/vikit_gateway.py:950]: Is this case ever used ? I mean why should we fallback implicitly to b64 when the provided file is not a plain image binary?

File: vikit/gateways/vikit_gateway.py [hunk 2]
```diff
@@ -1001,3 +1002,126 @@ async def generate_video_from_image_and_text_runway(self, prompt, aspect_ratio=(
         except Exception as e:
             logger.error(f"Error generating video from prompt: {e}")
             raise
+    
+    @retry(stop=stop_after_attempt(get_nb_retries_http_calls()), reraise=True)
+    async def ask_gemini(self, prompt, gemini_version="gemini-1.5-pro-002", more_contents = None):
```
Comment (jeffmac-aix) [inline vikit/gateways/vikit_gateway.py:1007]: Function comments would be useful here, like an explanation if what "more_contents" means

File: vikit/gateways/vikit_gateway.py [hunk 3]
```diff
@@ -1001,3 +1002,126 @@ async def generate_video_from_image_and_text_runway(self, prompt, aspect_ratio=(
...
+                "data": img_b64, 
+                "mimeType": "image/png"
+            }
+            parts_array.append(part)
+        
+        if prompt.video is not None and prompt.video.startswith("http"):
+            part={}
+            part["fileData"] = {
+                "fileUri": prompt.video,
+                "mimeType": "video/mp4"
```
Comment (jeffmac-aix) [inline vikit/gateways/vikit_gateway.py:1065]: is this always true? what if provided video is another format?

File: vikit/gateways/vikit_gateway.py [hunk 4]
```diff
@@ -1001,3 +1002,126 @@ async def generate_video_from_image_and_text_runway(self, prompt, aspect_ratio=(
...
+            }
+
+            parts_array.append(part)
+        elif prompt.video is not None:
+            part={}
+            video_data = ""
+            mimetype = "video/mp4"
```
Comment (jeffmac-aix) [inline vikit/gateways/vikit_gateway.py:1072]: same question as above, are we sure the provided video will be MP4 format?

File: vikit/gateways/vikit_gateway.py [hunk 5]
```diff
@@ -1001,3 +1002,126 @@ async def generate_video_from_image_and_text_runway(self, prompt, aspect_ratio=(
...
+            #We check if the video is a local path
+            if prompt.video.split('.') is not None and prompt.video.split('.')[-1].lower() in ["mp4", "mov", "avi", "wmv", "webm"]:
+                #Read file path and then convert to Base64
+                with open(prompt.video, "rb") as video_file:
+                    video_data = base64.b64encode(video_file.read()).decode(
+                        "utf-8"
+                    )
+                    mimetype = "video/" + prompt.video.split(".")[-1].lower()
+            else: 
+                # Convert result to Base64
```
Comment (jeffmac-aix) [inline vikit/gateways/vikit_gateway.py:1083]: the comment says we convert to base64, it does not seems so, should we?

File: vikit/gateways/vikit_gateway.py [hunk 6]
```diff
@@ -1001,3 +1002,126 @@ async def generate_video_from_image_and_text_runway(self, prompt, aspect_ratio=(
...
+                "fileUri": prompt.image, 
+                "mimeType": "image/" + prompt.image.split(".")[-1]
+            }
+
+            parts_array.append(part)
+        elif prompt.image is not None:
+            part={}
+            
+            if prompt.image.split('.') is not None and prompt.image.split('.')[-1].lower() in ["jpg", "png", "jpeg", "bmp", "tiff", "webp"]:
+                #Read file path and then convert to Base64
```
Comment (jeffmac-aix) [inline vikit/gateways/vikit_gateway.py:1034]: I do not see the file being converted to b64 here, it happens on the else branch

File: vikit/gateways/vikit_gateway.py (hunk 1 above)
Comment (leclem) [reply vikit/gateways/vikit_gateway.py:950]: There are three ways to give an image : http, local file, or base64. We first test if startswith http, if not we test if it is an image, if not then it should be a base64

File: vikit/gateways/vikit_gateway.py (hunk 4 above)
Comment (leclem) [reply vikit/gateways/vikit_gateway.py:1072]: true

File: vikit/gateways/vikit_gateway.py (hunk 5 above)
Comment (leclem) [reply vikit/gateways/vikit_gateway.py:1083]: true it already is

File: vikit/gateways/vikit_gateway.py (hunk 3 above)
Comment (jeffmac-aix) [reply vikit/gateways/vikit_gateway.py]: lgtm

File: vikit/gateways/vikit_gateway.py (hunk 4 above)
Comment (jeffmac-aix) [reply vikit/gateways/vikit_gateway.py]: lgtm

File: vikit/gateways/vikit_gateway.py (hunk 5 above)
Comment (jeffmac-aix) [reply vikit/gateways/vikit_gateway.py]: lgtm

File: vikit/gateways/vikit_gateway.py (hunk 6 above)
Comment (jeffmac-aix) [reply vikit/gateways/vikit_gateway.py]: lgtm

**Pull request merged**